tests/queries/test_bulk_update.py

```diff
@@ -250,3 +251,23 @@ def test_json_field(self):
             obj.json_field = {'c': obj.json_field['a'] + 1}
         JSONFieldNullable.objects.bulk_update(objs, ['json_field'])
         self.assertCountEqual(JSONFieldNullable.objects.filter(json_field__has_key='c'), objs)
+
+    def test_nullable_fk_after_related_save(self):
+        parent = RelatedObject.objects.create()
+        child = SingleObject()
+        parent.single = child
+        parent.single.save()
+        RelatedObject.objects.bulk_update([parent], fields=['single'])
+        self.assertEqual(parent.single_id, parent.single.pk)
+        parent.refresh_from_db()
+        self.assertEqual(parent.single, child)
+
+    def test_unsaved_parent(self):
+        parent = RelatedObject.objects.create()
+        parent.single = SingleObject()
+        msg = (
+            "bulk_update() prohibited to prevent data loss due to unsaved "
+            "related object 'single'."
+        )
+        with self.assertRaisesMessage(ValueError, msg):
+            RelatedObject.objects.bulk_update([parent], fields=['single'])
```

Comment:
Wonder if we should add a optional `fields` argument to `_prepare_related_fields_for_save` to only run the logic on the provided fields. Otherwise this exception will be raised for fields not specified through the `fields` argument on `bulk_update`.

Reply:
Good catch, I think it might be worth doing and testing for yes.